src/Land/Model/SoilWaterParameterizations.jl

```diff
@@ -0,0 +1,512 @@
+"""
+struct NoImpedance{FT} <: AbstractImpedanceFactor{FT} end
+
+
+
+"""
+struct IceImpedance{FT} <: AbstractImpedanceFactor{FT}
+
+The necessary parameters for the empirical impedance factor due to ice.
+
+# Fields
+$(DocStringExtensions.FIELDS)
+
+"""
+Base.@kwdef struct IceImpedance{FT} <: AbstractImpedanceFactor{FT}
+    "Empirical coefficient from Hansson 2014. "
+    Ω::FT = FT(7)
+end
+
+"""
+function impedance_factor(
+    imp::NoImpedance{FT},
+    θ_ice::FT,
+    porosity::FT,
+) where {FT}
+
+Returns the impedance factor when no effect due to ice is desired. Returns 1.
+
+The other arguments are included to unify the function call.
+"""
+function impedance_factor(
+    imp::NoImpedance{FT},
+    θ_ice::FT,
+    porosity::FT,
+) where {FT}
+    gamma = FT(1.0)
+    return gamma
+end
+
+"""
+function impedance_factor(
+    imp::IceImpedance{FT},
+    θ_ice::FT,
+    porosity::FT,
+) where {FT}
+
+Returns the impedance factor when an effect due to the fraction of ice is desired. 
+
+"""
+function impedance_factor(
+    imp::IceImpedance{FT},
+    θ_ice::FT,
+    porosity::FT,
+) where {FT}
+    Ω = imp.Ω
+    S_ice = θ_ice / porosity
+    gamma = FT(10.0^(-Ω * S_ice))
+    return gamma
+end
+
+"""
+function hydraulic_conductivity(
+    impedance::AbstractImpedanceFactor{FT},
+    viscosity::AbstractViscosityFactor{FT},
+    moisture::AbstractMoistureFactor{FT},
+    hydraulics::AbstractHydraulicsModel{FT},
+    θ_ice::FT,
+    porosity::FT,
+    T::FT,
+    S_l::FT,
+) where {FT}
+
+Returns the hydraulic conductivity.
+
+
+"""
+function hydraulic_conductivity(
+    impedance::AbstractImpedanceFactor{FT},
+    viscosity::AbstractViscosityFactor{FT},
+    moisture::AbstractMoistureFactor{FT},
+    hydraulics::AbstractHydraulicsModel{FT},
+    θ_ice::FT,
+    porosity::FT,
+    T::FT,
+    S_l::FT,
+) where {FT}
+    K = FT(
+        viscosity_factor(viscosity, T) *
+        impedance_factor(impedance, θ_ice, porosity) *
+        moisture_factor(moisture, hydraulics, S_l),
+    )
+    return K
+end
+
+"""
+    
+    hydraulic_head(z,ψ)
+
+Return the hydraulic head.
+
+The hydraulic head is defined as the sum of vertical height z and pressure head ψ; meters.
+"""
+hydraulic_head(z, ψ) = z + ψ
+
+
+
+"""
+   effective_saturation(porosity::FT, θ_l::FT)
+
+Compute the effective saturation of soil.
+
+
+θ_l is defined to be zero or positive. If θ_l is negative, hydraulic functions that take it as an argument will return imaginary numbers, resulting in domain errors. However, it is possible that our current solver returns a negative θ_l due to numerical issues. Provide a warning in this case, and correct the value of θ_l so that the integration can proceed. We will remove this once the numerical issues are resolved.
+
+"""
+function effective_saturation(porosity::FT, θ_l::FT) where {FT}
+
+    if θ_l < 0
+        @show θ_l
+        @warn("Augmented liquid fraction is negative - domain error. Artificially setting equal to zero to proceed. ")
```

Comment:
I don't think we can leave the `@show` and `@warn` here because it's not GPU compatible and it will result in very large print logs (this function will be called at every nodal point for every timestep). Perhaps this should throw an error. See [here](https://github.com/CliMA/ClimateMachine.jl/blob/master/src/Common/Thermodynamics/relations.jl#L1107-L1127) how print/error debugging is done in Thermodynamics.jl. `error_on_non_convergence` is defined in `Thermodynamics.jl`.

Reply:
yeah - I actually think that we dont need to make a version with an error_on_theta_negative attribute that either sets theta to zero when negative (if False) or gives an error when theta is negative (if True). If the user wants to set set it to zero in this case for debugging, they can do so, but we dont need to have clean code in master that allows for debugging. @cfranken @charleskawczynski  I guess I am proposing only this:
```
function effective_saturation(porosity::FT, θ_l::FT) where {FT}

    if θ_l < 0
        @print("-----------------------------------------\n")
        @print("Effective saturation is negative:\n")
        @print("θ_l =",
                θ_l,
               "\n"
               )
        error("Effective saturation is negative. Exiting...\n")
    end
    S_l = θ_l / porosity
    return S_l
end
```

Reply:
(if someone wants to debug, they can make the change to debug, and then change it back when bug is fixed)

Reply:
Edit - going with throw(), because then I can test that the correct exception is thrown in unit tests:

```
function effective_saturation(porosity::FT, θ_l::FT) where {FT}

    if θ_l < 0
        throw(DomainError(θ_l, "Effective saturation is negative."))
    end
    S_l = θ_l / porosity
    return S_l
end
```

Reply:
You can use `test_throws ErrorException effective_saturation(...)`. Also, you'll need to add `using KernelAbstractions: @print` to the top of the module.

Reply:
ok awesome! did so. this doesnt need @print then, so i think we are good (the message is printed in the exception).

going to resolve, I also updated repo to reflect these changes + the last formatting issues you mentioned (line lengths, """ on line before functions)